### PR TITLE
Choose minutes

### PR DESCRIPTION
**Pull request by CasperLaiTW**
choose-minutes

### CODE REVIEW

File: lib/js/index.js [hunk 1]
```diff
@@ -126,12 +126,21 @@ class DateTimePicker extends Events {
     this.$('.js-ok')
       .addEventListener('click', () => this.clickSubmit(), false);
 
+    this.$('.c-datepicker__header-date__hours').addEventListener('click', e => this.showHourClock(e), false);
```
Comment (jwhitfieldseed) [inline lib/js/index.js]: Please use `.js-date-hours` here - BEM classnames are for style, `js-` classes are for JavaScript. I know we don't totally adhere to this yet, but that's the aim 😄 

The idea comes from [this article](https://csswizardry.com/2015/03/more-transparent-ui-code-with-namespaces/) by Harry Roberts, btw - it's a good read!

File: lib/js/index.js [hunk 2]
```diff
@@ -126,12 +126,21 @@ class DateTimePicker extends Events {
...
+    this.$('.c-datepicker__header-date__minutes').addEventListener('click', e => this.showMinuteClock(e), false);
```
Comment (jwhitfieldseed) [inline lib/js/index.js]: `.js-date-minutes`, please

File: lib/js/index.js [hunk 3]
```diff
@@ -126,12 +126,21 @@ class DateTimePicker extends Events {
...
+
     this.$('.c-datepicker__clock__hours').addEventListener('mouseleave', e => this.mouseOutClock(e), false);
-    this.$(`.${this.options.styles.clockNum}`).forEach((el) => {
+    this.$(`.c-datepicker__clock__hours .${this.options.styles.clockNum}`).forEach((el) => {
```
Comment (jwhitfieldseed) [inline lib/js/index.js]: Can you add a `.js-` class for this the `hours` element?

File: lib/js/index.js [hunk 4]
```diff
@@ -126,12 +126,21 @@ class DateTimePicker extends Events {
...
       el.addEventListener('click', e => this.clickClock(e), false);
       el.addEventListener('mouseenter', e => this.mouseInClock(e), false);
     });
 
+    this.$('.c-datepicker__clock__minutes').addEventListener('mouseleave', e => this.mouseOutClock(e), false);
```
Comment (jwhitfieldseed) [inline lib/js/index.js]: `.js-` class please. I'll stop making this comment now, but it should apply to each new className referenced in JavaScript.

I'm going to do the same to any pre-existing non `js-` classes soon.

File: lib/js/index.js [hunk 5]
```diff
@@ -171,6 +180,13 @@ class DateTimePicker extends Events {
     return this;
   }
 
+  clickClockMinutes(e) {
```
Comment (jwhitfieldseed) [inline lib/js/index.js]: (Github won't let me add a comment to the right line) - do you need to change `clickClock` to `clickClickHour`?

I changed `clickClock` recently to call `this.set` (to fix #110) - can you update this method to follow the same pattern?

File: lib/js/index.js [hunk 6]
```diff
@@ -216,6 +240,20 @@ class DateTimePicker extends Events {
     return this;
   }
 
+  showHourClock() {
+    this.$('.c-datepicker__clock__hours').style.display = 'block';
```
Comment (jwhitfieldseed) [inline lib/js/index.js]: It seems inconsistent to use `style.display` for one part of the UI and a className for the other. Could you use an `active` class for both?

**Comment by jwhitfieldseed** [comment]
This is awesome, thank you! One usability comment - the hour and minute have `cursor: pointer` when the calendar is active. They are not clickable in this state.

Two possible fixes:
* Only use `cursor: pointer` when clock is active
* When the calendar is active, clicking hours/minutes opens the clock

The second option would be cool, but maybe it's one for another PR. If we do this we should also allow clicking on the date to open the calendar.

What do you think?

**Comment by CasperLaiTW** [comment]
I fix requested changes :)

About click to open clock, I make first option for this PR.

And I think I can keep working for second option include clicking on the date to open the calendar

File: lib/js/index.js [hunk 7]
```diff
@@ -126,12 +126,21 @@ class DateTimePicker extends Events {
...
-    this.$('.c-datepicker__clock__hours').addEventListener('mouseleave', e => this.mouseOutClock(e), false);
-    this.$(`.${this.options.styles.clockNum}`).forEach((el) => {
-      el.addEventListener('click', e => this.clickClock(e), false);
+    this.$('.js-date-hours').addEventListener('click', e => this.showHourClock(e), false);
+    this.$('.js-date-minutes').addEventListener('click', e => this.showMinuteClock(e), false);
+
+    this.$('.js-clock-hours').addEventListener('mouseleave', e => this.mouseOutClock(e), false);
+    this.$(`.js-clock-hours .${this.options.styles.clockNum}`).forEach((el) => {
+      el.addEventListener('click', e => this.clickClickHour(e), false);
       el.addEventListener('mouseenter', e => this.mouseInClock(e), false);
```
Comment (jwhitfieldseed) [inline lib/js/index.js]: I missed this before - can you change to `mouseInClockHour` ?

File: lib/js/index.js [hunk 8]
```diff
@@ -188,8 +206,16 @@ class DateTimePicker extends Events {
...
 
+  mouseInMinuteClock() {
+    const active = this.$(`.c-datepicker__clock__minutes .${this.options.styles.clockNum}--active`);
+
+    if (active) {
+      active.classList.add('hide-hand');
+    }
+  }
+
   mouseOutClock() {
```
Comment (jwhitfieldseed) [inline lib/js/index.js]: Can you make this `mouseOutHourClock` ?

File: lib/js/index.js [hunk 9]
```diff
@@ -291,17 +331,30 @@ class DateTimePicker extends Events {
...
       oldActiveHours.classList.remove(`${this.options.styles.clockNum}--active`);
     }
+
+    if (oldActiveMinutes) {
+      oldActiveMinutes.classList.remove(`${this.options.styles.clockNum}--active`);
+    }
+
     this.$(`.c-datepicker__clock__hours .${this.options.styles.clockNum}[data-number="${hourAsInt}"]`)
       .classList.add(`${this.options.styles.clockNum}--active`);
+    this.$(`.c-datepicker__clock__minutes .${this.options.styles.clockNum}[data-number="${minuteAsInt}"]`)
```
Comment (jwhitfieldseed) [inline lib/js/index.js]: I found a bug in testing - if the minutes from the datepicker's `default` time are not on the clock faces, this throws:

```
index.js:353 Uncaught TypeError: Cannot read property 'classList' of undefined
    at DateTimePicker.setTime (index.js:353)
    at DateTimePicker.open (index.js:92)
    at HTMLAnchorElement.<anonymous>
```

You can test this by adding `default: moment()` to the Datepicker options. If the current minutes is not a multiple of 5, you'll hit this error.

File: lib/js/index.js [hunk 10]
```diff
@@ -188,8 +206,16 @@ class DateTimePicker extends Events {
     }
   }
 
+  mouseInMinuteClock() {
+    const active = this.$(`.c-datepicker__clock__minutes .${this.options.styles.clockNum}--active`);
```
Comment (jwhitfieldseed) [inline lib/js/index.js]: Can you use `.js-clock-minutes` instead of `.c-datepicker__clock__minutes`

File: lib/js/index.js [hunk 11]
```diff
@@ -216,6 +242,20 @@ class DateTimePicker extends Events {
     return this;
   }
 
+  showHourClock() {
+    this.$('.c-datepicker__clock__hours').classList.add('active');
```
Comment (jwhitfieldseed) [inline lib/js/index.js]: `.js-clock-hours`

File: lib/js/index.js [hunk 12]
```diff
@@ -216,6 +242,20 @@ class DateTimePicker extends Events {
...
+    this.$('.c-datepicker__clock__minutes').classList.remove('active');
```
Comment (jwhitfieldseed) [inline lib/js/index.js]: `js-clock-minutes`

File: lib/js/index.js [hunk 13]
```diff
@@ -216,6 +242,20 @@ class DateTimePicker extends Events {
...
+    this.$('.c-datepicker__header-date__hours').classList.add('active');
```
Comment (jwhitfieldseed) [inline lib/js/index.js]: `.js-date-hours`

File: lib/js/index.js [hunk 14]
```diff
@@ -216,6 +242,20 @@ class DateTimePicker extends Events {
...
+    this.$('.c-datepicker__header-date__minutes').classList.remove('active');
```
Comment (jwhitfieldseed) [inline lib/js/index.js]: `.js-date-minutes`

File: lib/js/index.js [hunk 15]
```diff
@@ -216,6 +242,20 @@ class DateTimePicker extends Events {
...
+  }
+
+  showMinuteClock() {
+    this.$('.c-datepicker__clock__hours').classList.remove('active');
```
Comment (jwhitfieldseed) [inline lib/js/index.js]: The same className changes as above, please

File: lib/js/index.js [hunk 16]
```diff
@@ -291,17 +331,30 @@ class DateTimePicker extends Events {
   setTime(time) {
     const m = moment(time);
     const hour = m.format('HH');
+    const minutes = m.format('mm');
     const hourAsInt = parseInt(hour, 10) % 12;
+    const minuteAsInt = parseInt(minutes, 10);
     const oldActiveHours = this.$(`.c-datepicker__clock__hours .${this.options.styles.clockNum}--active`);
```
Comment (jwhitfieldseed) [inline lib/js/index.js]: `.js-clock-hours`, please - this one was here already but this seems like a good time to improve the code quality while making similar changes 😄

File: lib/js/index.js [hunk 17]
```diff
@@ -291,17 +331,30 @@ class DateTimePicker extends Events {
...
+    const oldActiveMinutes = this.$(`.c-datepicker__clock__minutes .${this.options.styles.clockNum}--active`);
```
Comment (jwhitfieldseed) [inline lib/js/index.js]: `.js-clock-minutes`

File: lib/js/index.js (hunk 7 above)
Comment (CasperLaiTW) [inline lib/js/index.js]: do you mean `mouseInClock` change to `mouseInHourClock` ?

**Comment by CasperLaiTW** [comment]
I make changes
About bug I make it will go next multiple of 5

File: lib/js/index.js (hunk 7 above)
Comment (jwhitfieldseed) [inline lib/js/index.js]: Yes, that seems a better name. Thanks!

Comment (CasperLaiTW) [inline lib/js/index.js]: I make changes, can you check it?

**Comment by jwhitfieldseed** [comment]
Thanks for the contribution, this is such an important feature! Released as [v2.1.0](https://github.com/ripjar/material-datetime-picker/releases/tag/v2.1.0).

**Pull request merged**